### ISSUE TITLE
as_array() should take aliases into account 

### ISSUE THREAD

**Issue by ccazette** [jonathangeiger/kohana-jelly#49]
If you query something using aliases, or even using field name instead of column names, wouldn't you expect as_array() to use this alias or field name as the key in the assoc array, instead of the column name ? If we say that Jelly can totally abstract the database, then you shouldn't have to know the column names when using the array...
Perhaps the query should use AS in such cases... Just thinking and suggesting...

**Comment by jonathangeiger** [comment]
Are you talking about as_array() in Jelly_Collection? Because it does alias fields to columns. Is there a place where you've found it isn't doing it properly?

http://github.com/jonathangeiger/kohana-jelly/blob/master/classes/jelly/collection/core.php#L82

**Comment by ccazette** [comment]
Yes it does indeed, when you specify key and value as parameters, you can use aliases which will eventually get converted to column_names before being used by the query_builder as_array(). 
I am talking about the fact the in the resulting array, the assoc keys of the array are not field names nor aliases, but the column names only. When Jelly is supposed to abstract these column names, how are you supposed to use the resulting assoc array?
You can for instance select for display_name in your query and have displayName as key in your array, displayName being the column name.

**Comment by ccazette** [comment]
Not sure there's an easy way to sort this out. Queries would have to use AS when selecting, otherwise I can't think about a way to have aliases in your array...

**Comment by banks** [comment]
Or we could modify the as_array method to go and manually rename the keys in the rows... possibly a big performance hit. I'd of thought AS was probably the way to go. This will take a bit of thought...

**Comment by jonathangeiger** [comment]
The problem with using AS is that you can no longer map columns to multiple fields. Currently, if there are two fields in a model that reference the same column they'll both get the value. If you use AS there has to be a 1:1 mapping of columns to fields.

If you're returning as an array, you probably have some data you're retrieving that doesn't really fit into any model. The easiest workaround for now is to just manual select and alias your columns.

```
Jelly::select('model')
    ->select(array('display_name', 'display_name');
```

Redundant? Yes.

**Comment by ccazette** [comment]
Hmmm... Sorry, jon, I just can't think of a situation where you would map the same column to different fields. Could you give an example ?

**Comment by jonathangeiger** [comment]
BelongsTo is a perfect example. Suppose you want a field to generate the related object but you also want direct access to the actual id in the column.

**Comment by ccazette** [comment]
Oh, right !
What about using AS anyway, if you have 2 fields mapping to the same column, then that would generate SELECT 'column' AS alias1, 'column' AS alias2 etc... ? The query would be slightly longer, but I suppose these situations are pretty rare... 
That's a fair bit of refactoring I have to admit... But well worth a thought ! :-)

**Comment by ccazette** [comment]
Hmm... Or you can still use ->execute(), and then use the Jelly_Model as_array() for every result. But that would be pretty slow...

**Comment by jonathangeiger** [comment]
> What about using AS anyway, if you have 2 fields mapping to the same column, then that would generate SELECT 'column' AS alias1, 'column' AS alias2 etc... 

This is a feasible solution. I wasn't even aware you could do that.

> Hmm... Or you can still use ->execute(), and then use the Jelly_Model as_array() for every result. But that would be pretty slow...

Jelly actually performs better than Sprig and ORM in this area. Jelly_Collection only instantiates one model per result and then loads values into it on each iteration. This means in Jelly's case you only instantiate one model object for any number of rows retrieved. With ORM you're loading 1 object for each row. With Sprig you're loading 1 object plus its field objects for each row.

**Comment by ccazette** [comment]
Interesting... I did not notice it ! That's great...
Iterating and use Jelly_Model as_array() is a solution then (still a bit slower than using as_array directly I believe, but acceptable). There is no solution is you want stdClasses however. 
I suppose that's you decision which way to go, use AS or leave it as it is ? I don't have a strong opinion, but I think AS makes sense.

**Comment by banks** [comment]
Yes the AS solution I think is the right one - one AS for each alias.

> Jelly actually performs better than Sprig and ORM in this area

To clarify: Jelly should use (roughly) the SAME memory as ORM when using MySQL driver since Database_Result_Mysql uses mysql_fetch_object() which means that there is only one instance of the object in memory at any one time.

Jelly_Collection is however quicker than ORM and Sprig's native use of Database_Result_Mysql since the overhead of instantiating a new object for each row is actually surprisingly large (all things relative here!). Since the object is not instantiated - only modified - for each iteration in a Jelly_Collection, it does take a little less time than iterating through the same database_result using as_object(). In all but the biggest result this is negligible though. Note a side effect is that in iterations, the object you get back each time is actually the SAME INSTANCE, just with all it's internal values reset to reflect the different model values.

**Comment by jonathangeiger** [comment]
> Note a side effect is that in iterations, the object you get back each time is actually the SAME INSTANCE, just with all it's internal values reset to reflect the different model values.

I guess that would cause problems only if you're comparing two models from the result set. I'll have to check and see if it's a problem in practice.

**Comment by jonathangeiger** [comment]
Also, just wanted to say I'll be implementing this rather soon. It needs to be fixed before we move forward with some of our other feature requests.

**Comment by jonathangeiger** [comment]
Pushed to unstable here: 6a8c0c056411e729aa8e8cdb8ceea0687e47bb52 and 419dcd38968846508c471d58358d1b8d071b85d8

Will have to add tests for it before we merge to master. It's a behavior change but I don't think it will affect many people.

Perhaps we should just wait until 0.9.6 to put it out?

**Comment by banks** [comment]
My bug fix in master may have clashes with this. I slightly changed the way \* gets aliased to get around some of the unpredictable issues. In master now \* always means \* and never gets a table added even when there are joins. If you want a specific model you must manually specify 'post._' other wise behaviour is unpredictable. I had to change with() to add the model name to the select(_) call to make this compatible. 

I'm not sure what the ramifications are for your changes there but I'm pretty sure this is the right way to go.

I'd say leave this for 0.9.6 - hopefully that is only a few days off anyway.

**Comment by jonathangeiger** [comment]
I looked over it. There might be a bit of clashing but it should be alright.

**Comment by ccazette** [comment]
Just wondering though.. What if 2 models have the same field name/alias, and you use with() ?

**Comment by banks** [comment]
@ccazzete: I wouldn't expect as_array() to return related values even if they are loaded using with(). Even if it did it would be as a multidimensional array so I can't see how this would cause a problem. Perhaps I misunderstand you.

**Comment by ccazette** [comment]
@banks : I am more talking about Jelly_Collection::as_array() in this case. I understand this is not a problem once you have a Model loaded as you already know which field is in which model. But when you have a flat db row, if you have joined two model having same aliases and use AS, I wonder how you sort this out.

Perhaps you already have, I am just curious as I came accross this this morning...

**Comment by jonathangeiger** [comment]
This has been closed by 5b3ee9e5e553926a28b9b240e3d62ad9c4beea37.
